File: ios/PTimerKit/Sources/PTimerCore/Exposure/NDFilterStack.swift
// Copyright © 2026 Sangwook Han
// SPDX-License-Identifier: Apache-2.0

import Foundation

/// A stack of one to four standard ND filter wheel values
/// (PTIMER-199). The stack is the domain shape behind the main
/// screen's ND wheel row: every entry is one wheel's committed value
/// in display order, and the whole stack collapses to a single
/// effective `NDStep` (the sum, in canonical stops) that feeds the
/// existing exposure calculation unchanged.
///
/// The 30-stop total limit is a DOMAIN INVARIANT of this type: every
/// construction and mutation boundary guarantees `sum ≤ 30`. The UI
/// additionally makes over-limit selections unrepresentable (wheels
/// select from ladders truncated to the remaining budget —
/// `remainingBudget(excludingWheelAt:)` +
/// `ExposureScale.ndSteps(upToStops:)`), but non-UI callers are held
/// to the same rule: `init` treats a violating stack as a programmer
/// error, and `replacingWheel(at:with:)` ignores a write that would
/// push the sum over the cap. Raw persisted values must be validated
/// BEFORE constructing a stack (the restore path rejects, it never
/// clamps).
public struct NDFilterStack: Equatable, Sendable {
    /// Maximum number of stacked filter wheels.
    public static let maximumWheelCount = 4

    /// One committed `NDStep` per wheel, in display order (1–4).
    public private(set) var entries: [NDStep]

    public init(entries: [NDStep]) {
        precondition(
            (1...Self.maximumWheelCount).contains(entries.count),
            "An ND filter stack holds 1...\(Self.maximumWheelCount) wheels."
        )
        precondition(
            entries.allSatisfy { $0.stops >= 0 },
            "ND filter wheels never hold negative stops."
        )
        precondition(
            Self.isWithinTotalLimit(entries),
            "An ND filter stack never exceeds \(ExposureScale.maximumWholeNDStops) stops in total."
        )
        self.entries = entries
    }

    /// Whether `entries` respect the 30-stop total limit (within the
    /// engine's stability epsilon). Exposed so validation layers
    /// (persistence restore) can check RAW values before attempting
    /// construction.
    public static func isWithinTotalLimit(_ entries: [NDStep]) -> Bool {
        entries.reduce(0) { $0 + $1.stops }
            <= Double(ExposureScale.maximumWholeNDStops)
                + ExposureCalculator.stabilityEpsilon
    }

    /// Single-wheel stack — the single-filter workflow's shape.
    public init(single step: NDStep = NDStep(stops: 0)) {
        self.init(entries: [step])
    }

    /// The one effective filter value the calculation consumes: the
    /// sum of every wheel in canonical stops.
    public var effectiveStep: NDStep {
        NDStep(stops: entries.reduce(0) { $0 + $1.stops })
    }

    /// Remaining stop budget available to the wheel at `index` under
    /// the 30-stop total limit: the cap minus every OTHER wheel's
    /// committed value. Feeds `ExposureScale.ndSteps(upToStops:)` to
    /// truncate that wheel's ladder.
    public func remainingBudget(excludingWheelAt index: Int) -> Double {
        precondition(entries.indices.contains(index), "Wheel index out of range.")
        let others = entries.enumerated()
            .filter { $0.offset != index }
            .reduce(0.0) { $0 + $1.element.stops }
        return Double(ExposureScale.maximumWholeNDStops) - others
    }

    // MARK: Mutations (all value-semantic)

    /// Whether another wheel can be added.
    public var canAddWheel: Bool {
        entries.count < Self.maximumWheelCount
    }

    /// Whether a wheel can be removed: more than one wheel AND at
    /// least one 0-stop wheel (wheels holding a value are never
    /// removed).
    public var canRemoveEmptyWheel: Bool {
        entries.count > 1 && entries.contains { $0.stops == 0 }
    }

    /// Appends one 0-stop wheel at the right; no-op at the maximum.
    /// A 0-stop wheel never changes `effectiveStep`.
    public func addingWheel() -> NDFilterStack {
        guard canAddWheel else {
            return self
        }
        var copy = self
        copy.entries.append(NDStep(stops: 0))
        return copy
    }

    /// Removes the 0-stop wheel at `index`; no-op for out-of-range
    /// indices, non-zero wheels, and single-wheel stacks. The
    /// overscroll gesture removes exactly the wheel it acted on
    /// (PTIMER-199 §4.2.3) — identity matters to the UI's reorder
    /// and removal animations even though zeros are value-identical.
    public func removingEmptyWheel(at index: Int) -> NDFilterStack {
        guard entries.indices.contains(index),
              entries[index].stops == 0,
              entries.count > 1 else {
            return self
        }
        var copy = self
        copy.entries.remove(at: index)
        return copy
    }

    /// Removes the rightmost 0-stop wheel; no-op when unavailable.
    public func removingRightmostEmptyWheel() -> NDFilterStack {
        guard canRemoveEmptyWheel,
              let index = entries.lastIndex(where: { $0.stops == 0 }) else {
            return self
        }
        var copy = self
        copy.entries.remove(at: index)
        return copy
    }

    /// Replaces one wheel's value. Out-of-range indices, negative or
    /// non-finite values, and writes that would push the total over
    /// the 30-stop limit are all ignored (defensive: the UI's
    /// budget-truncated ladders never produce any of these, so they
    /// are programmer-error shields for non-UI callers — every
    /// mutation boundary upholds the same invariants as `init`).
    public func replacingWheel(at index: Int, with step: NDStep) -> NDFilterStack {
        guard entries.indices.contains(index),
              step.stops >= 0,
              step.stops.isFinite else {
            return self
        }
        var copy = self
        copy.entries[index] = step
        guard Self.isWithinTotalLimit(copy.entries) else {
            return self
        }
        return copy
    }

    /// The post-commit ordering (agreed sort rule): descending by
    /// stops, 0-stop wheels rightmost, equal values keeping their
    /// existing relative order (stable). Never applied while a wheel
    /// is being scrolled; the caller triggers it on interaction end.
    /// Sorting reorders VALUES only, so `effectiveStep` is unchanged.
    public func sortedForCommit() -> NDFilterStack {
        var copy = self
        copy.entries = commitSortPermutation().map { entries[$0] }
        return copy
    }

    /// The index permutation `sortedForCommit()` applies:
    /// `permutation[i]` is the CURRENT index of the wheel that lands
    /// at position `i`. Exposed so a caller tracking per-wheel
    /// identity (for the reorder animation) can move companion state
    /// through the exact same stable sort.
    public func commitSortPermutation() -> [Int] {
        entries.enumerated()
            .sorted { lhs, rhs in
                if lhs.element.stops != rhs.element.stops {
                    return lhs.element.stops > rhs.element.stops
                }
                return lhs.offset < rhs.offset
            }
            .map(\.offset)
    }
}

extension ExposureScale {
    /// The scale's ND ladder truncated FROM THE TOP to `maxStops`
    /// (PTIMER-199 budget truncation). Top-truncation preserves the
    /// remaining rows' indices, so a wheel whose ladder shrinks keeps
    /// its selected row without jumping. Fractional presets above the
    /// budget drop out naturally.
    public func ndSteps(upToStops maxStops: Double) -> [NDStep] {
        ndSteps.filter { $0.stops <= maxStops + ExposureCalculator.stabilityEpsilon }
    }
}

File: android/core/src/main/kotlin/com/sangwook/ptimer/core/exposure/NdFilterStack.kt
// Copyright © 2026 Sangwook Han
// SPDX-License-Identifier: Apache-2.0

package com.sangwook.ptimer.core.exposure

import com.sangwook.ptimer.core.exposure.ExposureCalculator.Companion.STABILITY_EPSILON

/**
 * A stack of one to four standard ND filter wheel values (PTIMER-199).
 * Every entry is one wheel's committed value in display order, in
 * canonical stops; the whole stack collapses to a single effective
 * (summed) value that feeds the existing exposure calculation
 * unchanged. (iOS: NDFilterStack.)
 *
 * The 30-stop total limit is a DOMAIN INVARIANT of this type: every
 * construction and mutation boundary guarantees `sum <= 30`. The UI
 * additionally makes over-limit selections unrepresentable (wheels
 * select from ladders truncated to the remaining budget), but non-UI
 * callers are held to the same rule: construction rejects a violating
 * stack, and [replacingWheel] ignores a write that would push the sum
 * over the cap. Raw persisted values must be validated BEFORE
 * constructing a stack (the restore path rejects, it never clamps).
 */
data class NdFilterStack(val entries: List<Double>) {
    init {
        require(entries.size in 1..MAX_WHEEL_COUNT) {
            "An ND filter stack holds 1..$MAX_WHEEL_COUNT wheels."
        }
        require(entries.all { it >= 0.0 && it.isFinite() }) {
            "ND filter wheels never hold negative or non-finite stops."
        }
        require(isWithinTotalLimit(entries)) {
            "An ND filter stack never exceeds $MAX_TOTAL_STOPS stops in total."
        }
    }

    /** The one effective filter value the calculation consumes. */
    val effectiveStops: Double get() = entries.sum()

    /**
     * Remaining stop budget available to the wheel at [index] under the
     * 30-stop total limit: the cap minus every OTHER wheel's committed
     * value. Feeds the per-wheel ladder truncation.
     */
    fun remainingBudget(excludingWheelAt: Int): Double {
        require(excludingWheelAt in entries.indices) { "Wheel index out of range." }
        val others = entries.filterIndexed { i, _ -> i != excludingWheelAt }.sum()
        return MAX_TOTAL_STOPS - others
    }

    val canAddWheel: Boolean get() = entries.size < MAX_WHEEL_COUNT

    /**
     * Whether a wheel can be removed: more than one wheel AND at least
     * one 0-stop wheel (wheels holding a value are never removed).
     */
    val canRemoveEmptyWheel: Boolean
        get() = entries.size > 1 && entries.any { it == 0.0 }

    /** Appends one 0-stop wheel at the right; no-op at the maximum. */
    fun addingWheel(): NdFilterStack =
        if (canAddWheel) NdFilterStack(entries + 0.0) else this

    /** Removes the rightmost 0-stop wheel; no-op when unavailable. */
    fun removingRightmostEmptyWheel(): NdFilterStack {
        if (!canRemoveEmptyWheel) return this
        val index = entries.indexOfLast { it == 0.0 }
        if (index < 0) return this
        return NdFilterStack(entries.filterIndexed { i, _ -> i != index })
    }

    /**
     * Removes the 0-stop wheel at [index] — the overscroll gesture's
     * target (§4.2.3: exactly the wheel the photographer pulled).
     * No-op for out-of-range indices, non-zero wheels, and
     * single-wheel stacks.
     */
    fun removingEmptyWheel(at: Int): NdFilterStack {
        if (at !in entries.indices || entries[at] != 0.0 || entries.size <= 1) return this
        return NdFilterStack(entries.filterIndexed { i, _ -> i != at })
    }

    /**
     * Replaces one wheel's value. Out-of-range indices, negative or
     * non-finite values, and writes that would push the total over the
     * 30-stop limit are all ignored (reject, never clamp).
     */
    fun replacingWheel(at: Int, stops: Double): NdFilterStack {
        if (at !in entries.indices || stops < 0.0 || !stops.isFinite()) return this
        val replaced = entries.toMutableList().also { it[at] = stops }
        return if (isWithinTotalLimit(replaced)) NdFilterStack(replaced) else this
    }

    /**
     * The index permutation [sortedForCommit] applies: `result[i]` is
     * the CURRENT index of the wheel that lands at position `i`.
     * Exposed so a caller tracking per-wheel identity (the reorder
     * animation) can move companion state through the same stable
     * sort. Descending by stops, zeros rightmost, equal values keep
     * their existing relative order.
     */
    fun commitSortPermutation(): List<Int> =
        entries.withIndex()
            .sortedWith(compareByDescending<IndexedValue<Double>> { it.value }.thenBy { it.index })
            .map { it.index }

    /** The agreed post-commit ordering; reorders VALUES only. */
    fun sortedForCommit(): NdFilterStack =
        NdFilterStack(commitSortPermutation().map { entries[it] })

    companion object {
        const val MAX_WHEEL_COUNT: Int = 4
        const val MAX_TOTAL_STOPS: Double = 30.0

        /**
         * Whether [entries] respect the 30-stop total limit (within
         * the engine's stability epsilon). Exposed so validation
         * layers (persistence restore) can check RAW values before
         * attempting construction.
         */
        fun isWithinTotalLimit(entries: List<Double>): Boolean =
            entries.sum() <= MAX_TOTAL_STOPS + STABILITY_EPSILON

        /**
         * Restore validation (reject, never clamp): a persisted stack
         * is accepted only when the count, every value (finite,
         * non-negative, ON the shipping ladder), and the total pass —
         * otherwise the caller falls back to the legacy scalar.
         */
        fun isValidRestoredStack(entries: List<Double>?): Boolean {
            if (entries == null) return false
            if (entries.size !in 1..MAX_WHEEL_COUNT) return false
            if (!entries.all { it.isFinite() && it >= 0.0 }) return false
            if (!isWithinTotalLimit(entries)) return false
            return entries.all { value ->
                ExposureScale.shippingNDLadder.any { rung ->
                    kotlin.math.abs(rung.stops - value) <= STABILITY_EPSILON
                }
            }
        }
    }
}
